src/assets/js/components/popup/proposal.jsx

```diff
@@ -0,0 +1,386 @@
+                return undefined;
+              }
+
+              throw error;
+            });
+      })
+      .then((state) => {
+        currentProposal.state = state;
+
+        setAssetTitle(currentProposal.title);
+        setAssetUrl(currentProposal.url);
+        setAssetHostname(new URL(url).hostname);
+
+        setProposal(currentProposal);
+
+        if (state) {
+          setSaveAuthentication(state.withCookies);
+          setButtonState('update');
+        } else {
+          setButtonState('add');
+        }
+      })
+      .catch((error) => {
+        // Unknown error.
+        setError((prev) => {
+          return {
+            ...prev,
+            show: true,
+            message: 'Failed to check asset.'
+          };
+        });
+        throw error;
+      });
+  }
+
+  const proposeToAddToScreenly = async(user, url, title, cookieJar) => {
+    await updateProposal({
+      user,
+      title,
+      url: State.normalizeUrl(url),
+      cookieJar
+    });
+  };
+
+  const prepareToAddToScreenly = async() => {
+    const onlyPrimaryDomain = true;
+    const user = await getUser();
+
+    if (!user.token) {
+      return;
+    }
+
+    let tabs = await browser.tabs.query({ active: true, currentWindow: true });
+    let tabId = tabs[0].id;
+
+    try {
+      let result = await browser.scripting.executeScript({
+        target: { tabId: tabId },
+        func: () => {
+          return [
+            window.location.href,
+            document.title,
+            performance.getEntriesByType('resource').map(e => e.name)
+          ];
+        }
+      });
+
+      let [pageUrl, pageTitle, resourceEntries] = result[0].result;
+
+      if (!resourceEntries) {
+        return;
+      }
+
+      const originDomain = new URL(pageUrl).host;
+
+      let results = await Promise.all(
+        resourceEntries.map(url =>
+          browser.cookies.getAll({ url })
+        )
+      );
+
+      let cookieJar = Array.from(
+        new Map(
+          results
+            .flat(1)
+            .map(cookie =>
+              [
+                JSON.stringify([cookie.domain, cookie.name]),
+                cookie
+              ]
+            )
+        ).values()
+      );
+
+      if (onlyPrimaryDomain) {
+        // noinspection JSUnresolvedVariable
+        cookieJar = cookieJar.filter(cookie =>
+          cookie.domain === originDomain || (!cookie.hostOnly && originDomain.endsWith(cookie.domain))
+        )
+      }
+
+      await proposeToAddToScreenly(user, pageUrl, pageTitle, cookieJar);
+    } catch {
+      window.close();
+    }
+  };
+
+  useEffect(() => {
+    setIsLoading(true);
+    prepareToAddToScreenly()
+      .then(() => {
+        setIsLoading(false);
+      });
+  }, []);
+
+  const handleSubmission = async(event) => {
+    event.preventDefault();
+
+    let currentProposal = proposal;
+
+    if (buttonState === 'loading') {
+      return;
+    }
+
+    setButtonState('loading');
+    let headers = undefined;
```

Comment:
It will be undefined by default, even if you don't explicitly initialize to undefined

Reply:
A declaration is still needed. I can just do `let headers` or maybe `let headers = {}`